### ISSUE TITLE
Intermittent "System.Net.Http.HttpRequestException: No such host is known" when making high volume in-parallel calls from a single host.

### ISSUE THREAD

**Issue by veleek** [dotnet/runtime#31247]
This bug was originally filed against the Azure KeyVault client (here https://github.com/Azure/azure-sdk-for-net/issues/7873), but after some further troubleshooting it looks like it might be an issue related to the .NET framework.  Details are mostly copied from the initial bug report there, but please see that thread for more detail.  

There is a minimum repro of the issue (that I've been able to re-create) available here: https://gist.github.com/veleek/6751f70081e13d5f8bdf3f0cb09d0a4f and as mentioned below this only seems to occur with the Visual Studio debugger attached (which slows the process down).

**Describe the bug**
When making large numbers of requests to KeyVault, .NET occasionally throws 'No such host is known' errors.  Retrying with the exact same URL always succeeds 

In this case, I'm calling a KeyVault to enumerate all the secrets and then making calls in parallel to download them all.  I can consistently enumerate a few thousand secrets in a few minutes without issues, but then it starts slowing down, and then a few of these exceptions start occurring.  The KeyVault I'm testing with has 10030 secrets, and it successfully downloads ~9990 of them.  So it's a significant number of exception hits.

KeyVault requests are performed in parallel using `System.Threading.DataFlows.TransformBlock` with 25-50 requests in parallel.  

* This only seems to occur when running under the debugger in Visual Studio
* Visual Studio 2019
* .NET Core 3.0.100
* The problem doesn't reproduce 100%, it is possible to get a completely clean run but the repro rate is high. 
* When one of these errors occur, it almost always occurs a whole bunch of times for the remainder of the run.
* This is NOT related to service side throttling at all.  If that were the case then it would be more likely to occur when running without the debugger because the process completes all 10K keys MUCH faster and has never failed without the debugger.  I think the debugger may just be exposing a much more rare threading issue.

***Exception or Stack Trace***
```
System.Net.Http.HttpRequestException: No such host is known.
 ---> System.Net.Sockets.SocketException (11001): No such host is known.
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Rest.RetryDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.KeyVault.KeyVaultClient.GetSecretWithHttpMessagesAsync(String vaultBaseUrl, String secretName, String secretVersion, Dictionary`2 customHeaders, CancellationToken cancellationToken)
   at Microsoft.Azure.KeyVault.KeyVaultClientExtensions.GetSecretAsync(IKeyVaultClient operations, String secretIdentifier, CancellationToken cancellationToken)
   at ContentKey.Cli.KeyVaultContentKeySource.LoadKey(String keyId) in E:\src\ContentKey.Cli\KeyVaultContentKeySource.cs:line 70
   at ContentKey.Cli.IContentKeySource.<>c__DisplayClass2_0.<<UpdateKeys>b__0>d.MoveNext() in E:\src\ContentKey.Cli\IContentKeySource.cs:line 49
```

**Expected behavior**
No error should be thrown.

**Setup (please complete the following information):**
 - OS: Windows (.NET Core 3.0 Console App)
 - IDE : Visual Studio

**Additional context**
I'm able to reproduce this behavior pretty consistently when running with the debugger attached (no breakpoints configured or hit or anything though), but when running from the command line with no debugger it doesn't seem to fail.

**Comment by mikeball** [comment]
We have been seeing the same exception when making repeated http calls over the course of many hours to same host.

**Comment by davidsh** [comment]
>Additional context
I'm able to reproduce this behavior pretty consistently when running with the debugger attached (no breakpoints configured or hit or anything though), but when running from the command line with no debugger it doesn't seem to fail.

@veleek Are you still able to reproduce this problem? Can you reproduce it without running in the Visual Studio Debugger. It's possible the debugger is interfering with the high-load DNS and HTTP calls your app is making.

**Comment by karelz** [comment]
No response, closing.
If there is more information, feel free to reopen. Thanks!

**Comment by rapcgutz** [comment]
Any update regarding this issue? currently encounter this problem right now

**Comment by karelz** [comment]
@rapcgutz see the comments above -- we did not have enough data.
Please file a new issue and try to isolate the problem upfront to make it actionable for us.

**Comment by avatejo** [comment]
In my case the App throw the same error just if there is no internet connection.

**Comment by karelz** [comment]
@avatejo which is expected and should not be surprising, rigth?

**Comment by DonFrazier** [comment]
I am hitting this situation with .net core 3.1 in a console app.  I making a sustained high volume of POST requests to the same host/path but each request has a different body.  It appears retry succeeds.   The error usually occurs 10-20 minutes into a period of downloads. I have a somewhat straightforward code base that I can repro on a regular basis but it is not something I can post publicly.   PM me if you want me to try something in the code to get more info.

System.Net.Http.HttpRequestException, No such host is known.
System.Net.Http.HttpRequestException, The requested name is valid, but no data of the requested type was found.

This is all I am logging right now.  I often but not always see the 2nd message in conjunction withe the 1st.

**Comment by jorgy343** [comment]
I am getting the same error intermittently within an Azure function (.net core 3.1) when making ~1200 post calls per minute to a custom application hosted within our network. Unfortunately I can't provide a repro.

Has it been determined that this isn't an issue with .net or that the issue isn't consistent enough to determine a root cause?
